### PR TITLE
[Merged by Bors] - feat(NumberTheory/LSeries): functional equation for Dirichlet L-funcs

### PR DESCRIPTION
**Pull request by loefflerd**
Define completed L-functions for Dirichlet characters, prove that they are differentiable (except when they obviously aren't), and prove the functional equation.

---

This completes the (months-long) task of upstreaming the material from #10011.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/NumberTheory/LSeries/DirichletContinuation.lean [hunk 1]
```diff
@@ -10,22 +10,35 @@ import Mathlib.NumberTheory.DirichletCharacter.Basic
...
 * `LFunction_eq_LSeries`: if `1 < re s` then the `LFunction` coincides with the naive `LSeries`.
 * `differentiable_LFunction`: if `χ` is nontrivial then `LFunction χ s` is differentiable
   everywhere.
+* `LFunction_eq_completed_div_gammaFactor`: we have
+  `LFunction χ s = completedLFunction χ s / gammaFactor χ s`, unless `s = 0` and `χ` is the trivial
+  character modulo 1.
+* `differentiable_completedLFunction`: if `χ` is nontrivial then `completedLFunction χ s` is
+  differentiable everywhere.
+* `IsPrimitive.completedLFunction_one_sub`: if `χ` is primitive modulo `N`, then
+  `completedLFunction χ s = N ^ (s - 1 / 2) * rootNumber χ * completedLFunction χ⁻¹ s`.
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/DirichletContinuation.lean:36]: Mention "**functional equation**" somewhere here?

File: Mathlib/NumberTheory/LSeries/DirichletContinuation.lean [hunk 2]
```diff
@@ -67,4 +83,131 @@ lemma differentiable_LFunction {χ : DirichletCharacter ℂ N} (hχ : χ ≠ 1)
     Differentiable ℂ (LFunction χ) :=
   (differentiableAt_LFunction _ · <| Or.inr hχ)
 
+/-- The L-function of an even Dirichlet character vanishes at strictly negative even integers. -/
+@[simp]
+lemma Even.LFunction_neg_two_mul_nat_add_one {χ : DirichletCharacter ℂ N} (hχ : Even χ) (n : ℕ) :
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/DirichletContinuation.lean:88]: Maybe add a version for `-(2 * n)` with assumption `[NeZero n]`?

File: Mathlib/NumberTheory/LSeries/DirichletContinuation.lean [hunk 3]
```diff
@@ -67,4 +83,131 @@ lemma differentiable_LFunction {χ : DirichletCharacter ℂ N} (hχ : χ ≠ 1)
...
+@[pp_nodot] noncomputable def completedLFunction (χ : DirichletCharacter ℂ N) (s : ℂ) : ℂ :=
+  ZMod.completedLFunction χ s
+
+/--
+The completed L-function of the (unique) Dirichlet character mod 1 is the completed Riemann zeta
+function.
+-/
+lemma completedLFunction_modOne_eq {χ : DirichletCharacter ℂ 1} :
+    completedLFunction χ = completedRiemannZeta := by
+  ext1 s; rw [completedLFunction, ZMod.completedLFunction_modOne_eq, map_one, one_mul]
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/DirichletContinuation.lean:129]:
```suggestion
  ext1; rw [completedLFunction, ZMod.completedLFunction_modOne_eq, map_one, one_mul]
```

File: Mathlib/NumberTheory/LSeries/DirichletContinuation.lean [hunk 4]
```diff
@@ -67,4 +83,131 @@ lemma differentiable_LFunction {χ : DirichletCharacter ℂ N} (hχ : χ ≠ 1)
...
+
+/-- The root number of the unique Dirichlet character modulo 1 is 1. -/
+lemma rootNumber_modOne (χ : DirichletCharacter ℂ 1) : rootNumber χ = 1 := by
+  simp only [rootNumber, gaussSum, ← singleton_eq_univ (1 : ZMod 1), sum_singleton, map_one,
+    (show stdAddChar (1 : ZMod 1) = 1 from AddChar.map_zero_eq_one _), one_mul,
+    (show χ.Even from map_one _), ite_true, pow_zero, div_one, Nat.cast_one, one_cpow]
+
+namespace IsPrimitive
+
+/-- Functional equation for primitive Dirichlet L-functions. -/
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/DirichletContinuation.lean:178]: Maybe emphasise "The **Functional Equation** ..." ?

File: Mathlib/NumberTheory/LSeries/DirichletContinuation.lean [hunk 5]
```diff
@@ -67,4 +83,131 @@ lemma differentiable_LFunction {χ : DirichletCharacter ℂ N} (hχ : χ ≠ 1)
...
+  · simp only [completedLFunction_modOne_eq, completedRiemannZeta_one_sub, Nat.cast_one, one_cpow,
+      rootNumber_modOne, one_mul]
+  -- facts about `χ` as function
+  have h_sum : ∑ j, χ j = 0 := by
+    refine χ.sum_eq_zero_of_ne_one (fun h ↦ hN.symm ?_)
+    rwa [IsPrimitive, h, conductor_one (NeZero.ne _)] at hχ
+  let ε := I ^ (if χ.Even then 0 else 1)
+  -- gather up powers of N
+  rw [rootNumber, ← mul_div_assoc, ← div_mul_eq_mul_div₀, ← mul_div_assoc, ← div_mul_eq_mul_div₀,
+    ← cpow_sub _ _ (NeZero.ne _), sub_sub, add_halves]
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/DirichletContinuation.lean:192]:
```suggestion
  rw [rootNumber, ← mul_comm_div, ← mul_comm_div, ← cpow_sub _ _ (NeZero.ne _), sub_sub, add_halves]
```

**Review by MichaelStollBayreuth — APPROVED** [review]
Modulo a couple very minor things, this looks good to me.

**Comment by loefflerd** [comment]
Thanks for the review, Michael! I have applied all the suggested changes and merged master. Hope you are having a good time in Basel.

**Comment by MichaelStollBayreuth** [comment]
maintainer merge

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**